presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -302,49 +293,34 @@ protected Type visitSymbolReference(SymbolReference node, StackableAstVisitorCon
-            }
+            return handleResolvedField(node, scope.resolveField(node, node.getName()));
+        }
 
-            if (matches.size() > 1) {
-                throw new SemanticException(AMBIGUOUS_ATTRIBUTE, node, "Column '%s' is ambiguous", node.getName());
+        private Type handleResolvedField(Expression node, ResolvedField resolvedField)
+        {
+            if (scope != resolvedField.getScope()) {
+                throw new SemanticException(NOT_SUPPORTED, node, "Correlated queries not yet supported. Invalid column reference: '%s'", node);
             }
-
-            Field field = Iterables.getOnlyElement(matches);
-            int fieldIndex = tupleDescriptor.indexOf(field);
-            resolvedNames.put(node, fieldIndex);
-            expressionTypes.put(node, field.getType());
-            return field.getType();
+            columnReferences.add(node);
+            expressionTypes.put(node, resolvedField.getType());
+            return resolvedField.getType();
         }
 
         @Override
-        protected Type visitDereferenceExpression(DereferenceExpression node, StackableAstVisitorContext<AnalysisContext> context)
+        protected Type visitDereferenceExpression(DereferenceExpression node, StackableAstVisitorContext<Void> context)
         {
             QualifiedName qualifiedName = DereferenceExpression.getQualifiedName(node);
 
             // If this Dereference looks like column reference, try match it to column first.
             if (qualifiedName != null) {
-                List<Field> matches = tupleDescriptor.resolveFields(qualifiedName);
-                if (matches.size() > 1) {
-                    throw new SemanticException(AMBIGUOUS_ATTRIBUTE, node, "Column '%s' is ambiguous", node);
+                Optional<ResolvedField> resolvedField = scope.tryResolveField(node, qualifiedName);
+                if (resolvedField.isPresent()) {
+                    return handleResolvedField(node, resolvedField.get());
                 }
-
-                if (matches.size() == 1) {
-                    Field field = Iterables.getOnlyElement(matches);
-                    int fieldIndex = tupleDescriptor.indexOf(field);
-                    resolvedNames.put(node, fieldIndex);
-                    expressionTypes.put(node, field.getType());
-                    return field.getType();
-                }
-
-                if (!isColumnPrefix(qualifiedName, tupleDescriptor)) {
-                    if (isReferenceToOuterRelation(context.getContext(), qualifiedName)) {
-                        throw new SemanticException(NOT_SUPPORTED, node, "Correlated queries not yet supported. Invalid column reference: '%s'", qualifiedName);
-                    }
-                    throw createMissingAttributeException(node);
+                if (!scope.isColumnReference(qualifiedName)) {
+                    throw createMissingAttributeException(node, qualifiedName);
```

Comment:
Now I've seen the `SemanticExceptions` class and usage with `orElseThrow`. I'd consider migrating to `throwMissingAttributeException` and doing `optional.orElseThrow(() -> throwMissingAttributeException(...))` or  `optional.orElseGet(() -> throwMissingAttributeException(...))`, but I've got no strong preference to any of the ways (it's fine for me if it stays with `create...`)